### DISCUSSION TITLE
Implement CSRF token in NextJS

### DISCUSSION THREAD

**Discussion by umarmw** [vercel/next.js#13234] [Help]
Hello,

I'm currently using NextJS 9.3 and exporting plus serving my whole site statically (hosted on Azure Web app). According to Acunetix, all forms should have a CSRF token. Do you guys know how this can be achieved? 

Thanks,
Umar

**Comment by karesh** [comment]
Hi,

I am also having this question. I hope we can do something like this.

1. Create custom server.js (https://nextjs.org/docs/advanced-features/custom-server)
2. use csurf

```
var csrf = require('csurf')
app.use(csrf({ cookie: true }))
```
Any suggestion to this approach plz ?

**Reply by KempfCreative** [reply]
If you are serving your site statically, then this will not work.

**Comment by DesignerdriP** [comment]
If its server side, your for sure going to need them not because you won’t be safe without them really, but more security just doesn’t hurt anyone,  and you can label it as beneficial . Anyway they can be generated once per user session or for each request.. Anyhow per request  tokens are more secure than per-session ones.

**Comment by BleddP** [comment]
I have found a make-shift workaround to make this work in our app with a custom Express server. It's a bit of a makeshift solution but I hope this might benefit others in case they need it. 

Here's a simplified solution of how I implemented it.

**Server.js (custom express server)**
```
const express = require('express');
const next = require('next');
const url = require('url');
var csrf = require('csurf');
const cookieParser = require('cookie-parser');

// NextJS Configuration
const dev = process.env.NODE_ENV !== 'production';
const nextApp = next({ dev });
const handle = nextApp.getRequestHandler();

// Initiate the Express app
const PORT = process.env.PORT || 5000;
const app = express();

// CSRF protection middleware
var csrfProtection = csrf({ cookie: true });

// Initiate the NextApp
nextApp.prepare().then(() => {
  app.use(express.json());
  app.use(express.urlencoded({ extended: true }));
  app.use(cookieParser(process.env.COOKIE_PARSER_SECRET));

  // If you do not want your API routes protected with CSRF tokens, do not include the middlware
  app.use('/api/v1/wide-open', (req, res, next) => {
    return res.status(200).json({ message: 'This route is wide open' });
  });

  // If you want your API routes protected with CSRF
  app.use('/api/v1/protect-me', csrfProtection, (req, res, next) => {
    res.status(200).json({
      message: 'I am very safe',
    });
  });

  // Initialize CSRF to send a token to the front-end
  app.use(csrf({ cookie: true }));

  //catch-all for nextJS /pages
  app.get('*', (req, res) => {
    res.set({
      'Cache-Control': 'public, max-age=3600',
    });

    // It is important that the below two lines are inserted within the app.get('*') route
    const token = req.csrfToken();
    res.cookie('XSRF-TOKEN', token);

    const parsedUrl = url.parse(req.url, true);
    return handle(req, res, parsedUrl);
  });

  app.listen(PORT, (err) => {
    if (err) throw err;
    console.log('listening on port ' + PORT);
  });
});

```

We can then get the XSRF-TOKEN client side from document.cookie in _app.js 

**_app.js**

```
import React, { useEffect } from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import Head from 'next/head';

export default function MyApp(props) {
  const { Component, pageProps } = props;

  useEffect(() => {
    // Get the XSRF-TOKEN from cookies
    function getCookie(name) {
      const value = `; ${document.cookie}`;
      const parts = value.split(`; ${name}=`);
      if (parts.length === 2) return parts.pop().split(';').shift();
    }

    // set the 'csrf-token' as header on Axios POST requests only (please see csurf docs to see which other headers they accept)
    // you could also add PUT or PATCH if you wish
    axios.defaults.headers.post['csrf-token'] = getCookie('XSRF-TOKEN');

    // The rest of your UseEffect code (if any).....
  }, []);

  // Your app
  return (
    <React.Fragment>
      <Head></Head>
      <Navbar />
      <Component {...pageProps} />
      <Footer />
    </React.Fragment>
  );
}

MyApp.propTypes = {
  Component: PropTypes.elementType.isRequired,
  pageProps: PropTypes.object.isRequired,
};

```

The only thing I am unsure of is whether passing the token to the front-end has any security implications? I was always under the impression CSRF is handled server-side. However in the documentation for csurf they have examples for React where they pass it to either a req body or header. Maybe someone with more security insights can share their expertise?

Because we are not using sessions, the server generates two tokens, one called `_csrf` - this is normal, as this is the secret that csurf will validate against.

> **Notes**
> If you implement it this way and you are testing in Postman / Insomnia, a regular POST request will be rejected by the csurf middleware. So you'll first have to do a GET request to your website (or http://localhost:PORT in dev) and get the csrf token from the cookie it returns. This gets a bit annoying, so you can remove the middleware while you're in dev mode and make sure to add them before you go to production.

**Reply by younes-alouani** [reply]
I don't understand why this reply has 6 upvotes while it has two problems : 
- I need to change the next js server just to support csrf. Thus disabling all the features that a native next js server offers
- an attacker can embed javascript code that calls csrf api inside a form and send it to the victim. The victim will call csrf api using its cookies and thus making csrf implementation useless.

**Comment by Bluefitdev** [comment]
Does anyone manage to use this in production? I'm also very confused as how to implement this in nextjs

**Comment by francescovenica** [comment]
what about creating an handler like this:

```
import csurf from "csurf";
export default (req, res) =>
  new Promise((resolve, reject) => {
    return csurf({ cookie: true })(req, res, (error, res) => {
      if (error) reject(error);
      return resolve(res);
    });
  });
```

to be used in any api:
```
export default async (req, res) => {
  // protect api with csrf
  await csrf(req, res);
}
```

or page:
```
    const { req, res } = context;
    // protect api with csrf
    await csrf(req, res);  
    props = { ..., csrfToken: req.csrfToken() };
```

the token is passed as property so you can attach it to any ajax request of form or whatever...
I've just implemented this approach, I'm not sure yet if it is enough or if it cover all the edge cases...

**Reply by umer1989** [reply]
How to send csrf token at first time to next js page?

**Reply by francescovenica** [reply]
you can pass it using getServerSideProps and this:
```props = { ..., csrfToken: req.csrfToken() };```

I'm not 100% sure this implementation is the right way, it seems to work but still in development

**Comment by Izhaki** [comment]
https://github.com/j0lv3r4/next-csrf might help some.

**Reply by shane-arthur** [reply]
> https://github.com/j0lv3r4/next-csrf might help some.

This lib doesn't work with cookies and the maintainers aren't really responsive.

**Comment by toto1384** [comment]
How do you do it with next _middlewares?

**Reply by Trimud** [reply]
I wasn't able to make it work with _middleware. Unfortunately we don't have the Response there which should be passed as second arg to `await csrf(req, res);`

**Reply by toto1384** [reply]
Yeah, the best solution is still node-connect

**Reply by Trimud** [reply]
Meaning to have custom Express server to handle this?

**Reply by francescovenica** [reply]
why not the above solution? 
https://github.com/vercel/next.js/discussions/13234#discussioncomment-969390

this doesn't need any express server or node-connect

**Reply by toto1384** [reply]
1. @Trimud My bad, it isn't node-connect it's next-connect and it's a package that lets you add middlewares and still have access to the req and res ([Here is a link](https://www.npmjs.com/package/next-connect))
2. @francescovenica the solution works fine if you like to paste a line of code in every request you have, but most of us don't like that.
3. The [new next update](https://nextjs.org/blog/layouts-rfc) will fix this problem I think :))

**Reply by francescovenica** [reply]
probably I don't understand what you mean, I'm using this in a nextjs app with apollo and I'm adding the token only in one place for all the requests, so I'm not copy/paste anything on every request I have....if otherwise you mean that you need the token only for some particular requests then you can just whitelist requests that doesn't needs the csrf....
